### PR TITLE
[3.10] GH-93964: Harden overflow checks before _PyBytes_Resize in compile.c

### PR DESCRIPTION
**Pull request by sweeneyde**
https://github.com/python/cpython/issues/93964

### CODE REVIEW

**Comment by sweeneyde** [comment]
I confirmed that this successfully raises instead of crashing:
```
>>> import english_data  # from the issue
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
MemoryError
```

File: Python/compile.c [hunk 1]
```diff
@@ -6605,9 +6605,18 @@ static int
 assemble_emit_linetable_pair(struct assembler *a, int bdelta, int ldelta)
 {
     Py_ssize_t len = PyBytes_GET_SIZE(a->a_lnotab);
-    if (a->a_lnotab_off + 2 >= len) {
-        if (_PyBytes_Resize(&a->a_lnotab, len * 2) < 0)
+    if (a->a_lnotab_off > INT_MAX - 2) {
+        PyErr_NoMemory();
```
Comment (markshannon) [inline Python/compile.c]: Strictly this isn't a memory error, but a limitation of the VM.
Maybe raise an OverflowError?

(Likewise for the two cases below)

File: Python/compile.c [hunk 2]
```diff
@@ -6687,14 +6696,19 @@ assemble_emit(struct assembler *a, struct instr *i)
...
     size = instrsize(arg);
     if (i->i_lineno && !assemble_lnotab(a, i))
         return 0;
+    if (a->a_offset > INT_MAX - size) {
+        PyErr_NoMemory();
+        return 0;
+    }
     if (a->a_offset + size >= len / (int)sizeof(_Py_CODEUNIT)) {
-        if (len > PY_SSIZE_T_MAX / 2)
+        if (len > PY_SSIZE_T_MAX / 2) {
```
Comment (markshannon) [inline Python/compile.c]: This needs to be `INT_MAX / 2` not `PY_SSIZE_T_MAX / 2`.
We assume throughout the interpreter that various offsets and indexes fit into `int`.

**Review by markshannon — CHANGES_REQUESTED** [review]
A couple of minor issues, otherwise LGTM.

**Comment by markshannon** [comment]
Thanks.

**Pull request merged**